subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java

```diff
@@ -202,6 +215,10 @@ public String determineName(Configuration c) {
     /**
      * Takes a closure which gets coerced into a {@link Spec}. Behaves otherwise in the same way as
      * {@link #files(org.gradle.api.specs.Spec)}.
+     * 
+     * @implSpec Usage: This method should only be called on resolvable configurations and will fail if
+     * called on a configuration that does not permit this usage.  It should warn if called on a configuration that has
```

Comment:
I was thinking that using "should" in an implSpec communicated that implementers should really add these guards, but they will remain logically correct without them (hence not "must").  I think I'll update the "and will fail" to "and should fail" to be consistent with this.

I didn't get any confusion about deprecation - the implementations "should" warn if allowed but deprecated.  I'll update to "marked as allowed but is deprecated".

Reply:
This won't prevent then from knowing anything though.  Unless your saying they'll lack the confidence that we'll follow our own guidance here inside of the `DefaultConfiguration` implementation?  

They shouldn't be looking at `DefaultConfiguration` directly at all, they shouldn't even be aware of the name of that type.

I don't think I understand what you're suggesting to change here.

Reply:
I agree most users are not aware of DefaultConfiguration and should not need to be. The average user is going to be looking at the default `Configuration` documentation. I think the current wording of these implSpecs would leave users confused as to what the actual behavior of any given configuration is. In reality whatever `Configuration` object a user is going to get is going to be a `DefaultConfiguration`, and we can say with confidence here that these method will warn

Reply:
So you're suggesting change "should" to "will"?  

IDK, I think "should" is strong enough, and it's more correct to put this as part of an implSpec than as part of the "general" documentation on the intended usage of the method.  

If we said "will" (or "must") we're kind of changing the documented purpose of the interface method in a way I'd find more alarming as a client, rather than advising what implementers (yes, only us) should provide.

Reply:
> So you're suggesting change "should" to "will"?

Yup

> If we said "will" (or "must") we're kind of changing the documented purpose of the interface method in a way I'd find more alarming as a client, rather than advising what implementers (yes, only us) should provide.

We do this almost everywhere. There are plenty of places where we document the behavior of the actual implementation in the interface. I don't think we necessarily expose these interfaces because we expect any given user to be able to implement them -- but so that we are able to control a subset of an object's methods which we certify as stable for use in user code. If we used `should` everywhere I think it would leave our users much less confident in the actual behavior of the interfaces we expose. 

Leave this as you wish. I think "should" is confusing here.